File: src/domain/models/distribution.go
package models

type Hls struct {
	SegmentDuration int
	WindowSize      int // Number of segments in the live playlist (default: 3)
}

type Dash struct {
	SegmentDuration int
	WindowSize      int // Number of segments in the live manifest (default: 3)
}

type Distribution struct {
	Hls  *Hls
	Dash *Dash
}

func (d Distribution) HlsEnabled() bool {
	return d.Hls != nil
}

func (d Distribution) DashEnabled() bool {
	return d.Dash != nil
}

func (d Distribution) IsDualMode() bool {
	return d.HlsEnabled() && d.DashEnabled()
}

// SegmentDuration returns the segment duration from whichever format is configured.
// In dual mode, both must match (enforced by validation), so either is fine.
func (d Distribution) SegmentDuration() int {
	if d.Hls != nil {
		return d.Hls.SegmentDuration
	}
	if d.Dash != nil {
		return d.Dash.SegmentDuration
	}
	return 0
}

// WindowSize returns the window size from whichever format is configured.
func (d Distribution) WindowSize() int {
	if d.Hls != nil {
		return d.Hls.WindowSize
	}
	if d.Dash != nil {
		return d.Dash.WindowSize
	}
	return 3
}

File: src/main/java/com/fjourdren/theatrum/domain/model/Distribution.java
package com.fjourdren.theatrum.domain.model;

import com.fjourdren.theatrum.domain.constant.ConfigConstants;

/** Distribution formats configured for a stream. A null component means that format is disabled. */
public record Distribution(Hls hls, Dash dash) {

    public static Distribution none() {
        return new Distribution(null, null);
    }

    public static Distribution ofHls(Hls hls) {
        return new Distribution(hls, null);
    }

    public static Distribution ofDash(Dash dash) {
        return new Distribution(null, dash);
    }

    public boolean hlsEnabled() {
        return hls != null;
    }

    public boolean dashEnabled() {
        return dash != null;
    }

    public boolean isDualMode() {
        return hlsEnabled() && dashEnabled();
    }

    /**
     * Segment duration from whichever format is configured. In dual mode both must match
     * (enforced by config validation), so either is fine.
     */
    public int segmentDuration() {
        if (hls != null) {
            return hls.segmentDuration();
        }
        if (dash != null) {
            return dash.segmentDuration();
        }
        return 0;
    }

    /** Window size from whichever format is configured, defaulting to 3. */
    public int windowSize() {
        if (hls != null) {
            return hls.windowSize();
        }
        if (dash != null) {
            return dash.windowSize();
        }
        return ConfigConstants.DEFAULT_WINDOW_SIZE;
    }
}
